File: model/Task.py
import datetime
import enum
import uuid
from typing import Optional, TypeVar

from pydantic import BaseModel, Field
from pydantic.utils import to_camel

TaskId = TypeVar('TaskId', bound=str)


class TaskStatus(enum.Enum):
    READY = "READY"
    RUNNING = "RUNNING"
    PAUSED = "PAUSED"

    PENDING = "PENDING"

    CANCELED = "CANCELED"
    DETACHED = "DETACHED"
    ERROR = "ERROR"
    TERMINATED = "TERMINATED"


class ProcessInfo(BaseModel):
    current: int
    total: int

    class Config:
        alias_generator = to_camel
        allow_population_by_field_name = True


class TaskInfo(BaseModel):
    id: TaskId = Field(default_factory=lambda: str(uuid.uuid4()))

    status: TaskStatus = TaskStatus.PENDING
    process: Optional[ProcessInfo]

    created_at: datetime.datetime = datetime.datetime.now()
    started_at: Optional[datetime.datetime]
    paused_at: Optional[datetime.datetime]
    terminated_at: Optional[datetime.datetime]

    name: str
    description: str = ""

    working_dir: str = ""
    command_line: str
    output_file_path: str = "task.log"

    class Config:
        alias_generator = to_camel
        allow_population_by_field_name = True

File: pkg/model/task.go
package model

import (
	"github.com/google/uuid"
	pb "qtask/api/proto/v1"
	"time"
)

type TaskId string

type TaskStatus string

const (
	TaskStatusReady     TaskStatus = "READY"
	TaskStatusRunning   TaskStatus = "RUNNING"
	TaskStatusPaused    TaskStatus = "PAUSED"
	TaskStatusPending   TaskStatus = "PENDING"
	TaskStatusCanceled  TaskStatus = "CANCELED"
	TaskStatusCompleted TaskStatus = "COMPLETED"
	TaskStatusDetached  TaskStatus = "DETACHED"
	TaskStatusError     TaskStatus = "ERROR"
	TaskStatusNotExist  TaskStatus = "NOT_EXIST"
)

type TaskDetail struct {
	Id     TaskId
	Status TaskStatus

	CreatedAt    time.Time
	StartedAt    *time.Time
	PausedAt     *time.Time
	TerminatedAt *time.Time

	Name           string
	Description    string
	WorkingDir     string
	Path           string
	Args           []string
	OutputFilePath string
	ExitCode       int32
	ExitMessage    string
}

func NewTaskDetail(path string, args ...string) *TaskDetail {
	id := TaskId(uuid.New().String())
	createdAt := time.Now()
	return &TaskDetail{
		Id:     id,
		Status: TaskStatusPending,

		CreatedAt:    createdAt,
		StartedAt:    nil,
		PausedAt:     nil,
		TerminatedAt: nil,

		Name:           "unnamed",
		Description:    "",
		WorkingDir:     ".",
		Path:           path,
		Args:           args,
		OutputFilePath: "output.log",
		ExitCode:       0,
		ExitMessage:    "",
	}
}

func TaskDetailFromProto(proto *pb.TaskDetail) *TaskDetail {
	var startedAt *time.Time = nil
	var pausedAt *time.Time = nil
	var terminatedAt *time.Time = nil
	if proto.StartedAt != nil {
		t := proto.StartedAt.AsTime()
		startedAt = &t
	}
	if proto.PausedAt != nil {
		t := proto.PausedAt.AsTime()
		pausedAt = &t
	}
	if proto.TerminatedAt != nil {
		t := proto.TerminatedAt.AsTime()
		terminatedAt = &t
	}
	taskDetail := TaskDetail{
		Id:             TaskId(proto.TaskId),
		Status:         TaskStatus(proto.Status),
		CreatedAt:      proto.CreatedAt.AsTime(),
		StartedAt:      startedAt,
		PausedAt:       pausedAt,
		TerminatedAt:   terminatedAt,
		Name:           proto.Name,
		Description:    proto.Description,
		WorkingDir:     proto.WorkingDir,
		Path:           proto.Path,
		Args:           proto.Args,
		OutputFilePath: proto.OutputFilePath,
		ExitCode:       proto.ExitCode,
		ExitMessage:    proto.ExitMessage,
	}
	return &taskDetail
}
